src/ast/ast_build_ar_exp.c

```diff
@@ -0,0 +1,328 @@
+	AR_ExpNode *op = AR_EXP_NewOpNode((char *)func_name, arg_count);
+
+	for(unsigned int i = 0; i < arg_count; i ++) {
+		const cypher_astnode_t *arg = cypher_ast_apply_operator_get_argument(expr, i);
+		// Recursively convert arguments
+		op->op.children[i] = AR_EXP_FromExpression(record_map, arg);
+	}
+	return op;
+}
+
+static AR_ExpNode *_AR_EXP_FromIdentifierExpression(RecordMap *record_map,
+													const cypher_astnode_t *expr) {
+	// Identifier referencing another record_map entity
+	const char *alias = cypher_ast_identifier_get_name(expr);
+	return AR_EXP_NewVariableOperandNode(record_map, alias, NULL);
+}
+
+static AR_ExpNode *_AR_EXP_FromPropertyExpression(RecordMap *record_map,
+												  const cypher_astnode_t *expr) {
+	// Identifier and property pair
+	// Extract the entity alias from the property. Currently, the embedded
+	// expression should only refer to the IDENTIFIER type.
+	const cypher_astnode_t *prop_expr = cypher_ast_property_operator_get_expression(expr);
+	assert(cypher_astnode_type(prop_expr) == CYPHER_AST_IDENTIFIER);
+	const char *alias = cypher_ast_identifier_get_name(prop_expr);
+
+	// Extract the property name
+	const cypher_astnode_t *prop_name_node = cypher_ast_property_operator_get_prop_name(expr);
+	const char *prop_name = cypher_ast_prop_name_get_value(prop_name_node);
+
+	return AR_EXP_NewVariableOperandNode(record_map, alias, prop_name);
+}
+
+static AR_ExpNode *_AR_EXP_FromIntegerExpression(RecordMap *record_map,
+												 const cypher_astnode_t *expr) {
+	const char *value_str = cypher_ast_integer_get_valuestr(expr);
+	char *endptr = NULL;
+	int64_t l = strtol(value_str, &endptr, 0);
+	assert(endptr[0] == 0);
+	SIValue converted = SI_LongVal(l);
+	return AR_EXP_NewConstOperandNode(converted);
+}
+
+static AR_ExpNode *_AR_EXP_FromFloatExpression(RecordMap *record_map,
+											   const cypher_astnode_t *expr) {
+	const char *value_str = cypher_ast_float_get_valuestr(expr);
+	char *endptr = NULL;
+	double d = strtod(value_str, &endptr);
+	assert(endptr[0] == 0);
+	SIValue converted = SI_DoubleVal(d);
+	return AR_EXP_NewConstOperandNode(converted);
+}
+
+static AR_ExpNode *_AR_EXP_FromStringExpression(RecordMap *record_map,
+												const cypher_astnode_t *expr) {
+	const char *value_str = cypher_ast_string_get_value(expr);
+	SIValue converted = SI_ConstStringVal((char *)value_str);
+	return AR_EXP_NewConstOperandNode(converted);
+}
+
+static AR_ExpNode *_AR_EXP_FromTruExpression(RecordMap *record_map, const cypher_astnode_t *expr) {
+	SIValue converted = SI_BoolVal(true);
+	return AR_EXP_NewConstOperandNode(converted);
+}
+
+static AR_ExpNode *_AR_EXP_FromFalseExpression(RecordMap *record_map,
+											   const cypher_astnode_t *expr) {
+	SIValue converted = SI_BoolVal(false);
+	return AR_EXP_NewConstOperandNode(converted);
+}
+
+static AR_ExpNode *_AR_EXP_FromNullExpression(RecordMap *record_map, const cypher_astnode_t *expr) {
```

Comment:
I wouldn't bother passing the unused `record_map` to any of the functions that don't require it; it doesn't seem to simplify the interface to me.

Reply:
👍